sklearn/utils/validation.py

```diff
@@ -2151,7 +2151,11 @@ def _check_sample_weight(
         if force_float_dtype and dtype is None:
             dtype = float_dtypes
         if is_array_api and ensure_same_device:
-            sample_weight = xp.asarray(sample_weight, device=device)
+            dtype_param = {}
+            # We need to force `float32` with the `mps` device.
+            if str(device).startswith("mps"):
```

Comment:
I would rather avoid hardcoding device specific logic in the library code )(or at least limit such code to specific helper functions such as `_max_precision_float_dtype`).

Furthermore, it seems that this passing `force_float_dtype=True` should handle this case automatically for us.